File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/matlab/chapter5/svmhard.m
%% svmhard.m
% From A First Course in Machine Learning, Chapter 5.
% Simon Rogers, 01/11/11 [simon.rogers@glasgow.ac.uk]
% Hard margin SVM
clear all;close all;


%% Generate the data
x = [randn(20,2);randn(20,2)+4];
t = [repmat(-1,20,1);repmat(1,20,1)];

%% Plot the data
ma = {'ko','ks'};
fc = {[0 0 0],[1 1 1]};
tv = unique(t);
figure(1); hold off
for i = 1:length(tv)
    pos = find(t==tv(i));
    plot(x(pos,1),x(pos,2),ma{i},'markerfacecolor',fc{i});
    hold on
end

%% Setup the optimisation problem
N = size(x,1);
K = x*x';
H = (t*t').*K + 1e-5*eye(N);
f = repmat(1,N,1);
A = [];b = [];
LB = repmat(0,N,1); UB = repmat(inf,N,1);
Aeq = t';beq = 0;

% Following line runs the SVM
alpha = quadprog(H,-f,A,b,Aeq,beq,LB,UB);
% Compute the bias
fout = sum(repmat(alpha.*t,1,N).*K,1)';
pos = find(alpha>1e-6);
bias = mean(t(pos)-fout(pos));
%% Plot the data, decision boundary and Support vectors
figure(1);hold off
pos = find(alpha>1e-6);
plot(x(pos,1),x(pos,2),'ko','markersize',15,'markerfacecolor',[0.6 0.6 0.6],...
    'markeredgecolor',[0.6 0.6 0.6]);
hold on
for i = 1:length(tv)
    pos = find(t==tv(i));
    plot(x(pos,1),x(pos,2),ma{i},'markerfacecolor',fc{i});
end

xp = xlim;
% Because this is a linear SVM, we can compute w and plot the decision
% boundary exactly.
w = sum(repmat(alpha.*t,1,2).*x,1)';
yp = -(bias + w(1)*xp)/w(2);
plot(xp,yp,'k','linewidth',2)

File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/R/chapter5/svmhard.R
## svmhard.R
# From A First Course in Machine Learning, Chapter 5.
# Francois-Xavier Briol, 04/06/16 [f-x.briol@warwick.ac.uk]
# Hard margin SVM
rm(list=ls(all=TRUE))
require(quadprog) # might require to install this package: install.packages("quadprog")

## Generate the data
x = rbind(matrix(rnorm(2*20),ncol=2),matrix(rnorm(2*20)+4,ncol=2))
t = c(rep(-1,20),rep(1,20))

## Plot the data
tv = as.vector(unique(t))
ma = c(0,16)
col= c("blue","red")
pos = which(t==tv[1])
plot(x[pos,1],x[pos,2],col=col[1],pch=ma[1],xlim=c(-3,7),ylim=c(-3,7))
for(i in 1:length(tv)){
  pos = which(t==tv[i])
  points(x[pos,1],x[pos,2],col=col[i],pch=ma[i])
}

## Setup the SVM optimisation problem in the matrix form below
# $$ \min \frac{1}{2}b^T D b - d^T b$$
# s.t. $$ A^T b \geq b_0 $
# where the first inequality is treated as an equality.
N = nrow(x)
K = x%*%t(x)
D = (t%*%t(t))*K + (1e-5)*diag(N)
d = matrix(rep(1,N))
A = cbind(t,diag(N))
b0 = rep(0,N+1)
  
# Following line runs the SVM optimisation
alpha = solve.QP(Dmat=D,dvec=d,Amat=A,bvec=b0,meq=1)

# Compute the bias
fout = colSums(matrix(alpha$solution*t,ncol=N,nrow=N,byrow=FALSE)*K)
pos = which(alpha$solution>1e-3)
bias = mean(t[pos]-fout[pos])

##Plot the data, decision boundary and Support vectors
tv = as.vector(unique(t))
ma = c(0,16)
col= c("blue","red")
pos = which(t==tv[1])
plot(x[pos,1],x[pos,2],col=col[1],pch=ma[1],xlim=c(-3,7),ylim=c(-3,7))
for(i in 1:length(tv)){
  pos = which(t==tv[i])
  points(x[pos,1],x[pos,2],col=col[i],pch=ma[i])
}

# Because this is a linear SVM, we can compute w and plot the decision
# boundary exactly.
xp = c(-3,7)
w = colSums(matrix(alpha$solution*t,nrow=N,ncol=ncol(x),byrow=FALSE)*x)
yp = -(bias + w[1]*xp)/w[2]
points(xp,yp,type="l")




